hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieReaderFileIndex.scala

```diff
@@ -0,0 +1,118 @@
+import org.apache.spark.sql.execution.datasources.{FileStatusCache, NoopCache, PartitionDirectory}
+import org.apache.spark.sql.types.StructType
+
+class HoodieReaderFileIndex(override val spark: SparkSession,
+                            override val metaClient: HoodieTableMetaClient,
+                            override val schemaSpec: Option[StructType],
+                            override val options: Map[String, String],
+                            @transient override val fileStatusCache: FileStatusCache = NoopCache,
+                            override val includeLogFiles: Boolean = false,
+                            override val shouldEmbedFileSlices: Boolean = false)
+  extends HoodieFileIndex(
+    spark = spark,
+    metaClient = metaClient,
+    schemaSpec = schemaSpec,
+    options = options,
+    fileStatusCache = fileStatusCache,
+    includeLogFiles = includeLogFiles,
+    shouldEmbedFileSlices = shouldEmbedFileSlices) {
+
+  /**
+   *
+   * Returns set of indices with timestamp partition type. For Timestamp based keygen, there is only one
+   * partition so index is 0. For custom keygen, it is the partition indices for which partition type is
+   * timestamp.
+   */
+  private def getTimestampPartitionIndex(): Set[Int] = {
+    val tableConfig = metaClient.getTableConfig
+    val keyGeneratorClassNameOpt = Option.apply(tableConfig.getKeyGeneratorClassName)
+    val recordKeyFieldOpt = common.util.Option.ofNullable(tableConfig.getRawRecordKeyFieldProp)
+    val keyGeneratorClassName = keyGeneratorClassNameOpt.getOrElse(KeyGenUtils.inferKeyGeneratorType(recordKeyFieldOpt, tableConfig.getPartitionFieldProp).getClassName)
+    if (keyGeneratorClassName.equals(KeyGeneratorType.TIMESTAMP.getClassName)
+      || keyGeneratorClassName.equals(KeyGeneratorType.TIMESTAMP_AVRO.getClassName)) {
+      Set(0)
+    } else if (keyGeneratorClassName.equals(KeyGeneratorType.CUSTOM.getClassName)
+      || keyGeneratorClassName.equals(KeyGeneratorType.CUSTOM_AVRO.getClassName)) {
+      val partitionFields = HoodieTableConfig.getPartitionFieldsForKeyGenerator(tableConfig).orElse(java.util.Collections.emptyList[String]())
+      val partitionTypes = CustomAvroKeyGenerator.getPartitionTypes(partitionFields)
+      var partitionIndexes: Set[Int] = Set.empty
+      for (i <- 0 until partitionTypes.size()) {
+        if (partitionTypes.get(i).equals(PartitionKeyType.TIMESTAMP)) {
+          partitionIndexes = partitionIndexes + i
+        }
+      }
+      partitionIndexes
+    } else {
+      Set.empty
+    }
+  }
+
+  /**
+   * Invoked by Spark to fetch list of latest base files per partition.
+   *
+   * @param partitionFilters partition column filters
+   * @param dataFilters      data columns filters
+   * @return list of PartitionDirectory containing partition to base files mapping
+   */
+  override def listFiles(partitionFilters: Seq[Expression], dataFilters: Seq[Expression]): Seq[PartitionDirectory] = {
```

Comment:
Is this new index tested anywhere?

Reply:
Since this is the default file index for reading, it is being used by all the existing tests.